[SPARK-21440][SQL][PYSPARK] Refactor ArrowConverters and add ArrayType and StructType support.

ueshin opened the pull request:
## What changes were proposed in this pull request?

This is a refactoring of `ArrowConverters` and related classes.

1. Refactor `ColumnWriter` as `ArrowWriter`.
2. Add `ArrayType` and `StructType` support.
3. Refactor `ArrowConverters` to skip intermediate `ArrowRecordBatch` creation.

## How was this patch tested?

Added some tests and existing tests.

ueshin committed:
Introduce ArrowWriter and ArrowColumnVector.

ueshin committed:
Use ArrowWriter for ArrowConverters.

ueshin committed:
Refactor ArrowConverters.

ueshin committed:
Move releasing memory into task completion listener.

ueshin commented:
cc @cloud-fan @BryanCutler

SparkQA commented:
**[Test build #79668 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79668/testReport)** for PR 18655 at commit [`58cd465`](https://github.com/apache/spark/commit/58cd46506b02800269380f7c8acb5f9825664cad).

kiszk commented on sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ArrowColumnVector.java:
```diff
@@ -0,0 +1,502 @@
+    assert(dictionary == null);
+    boolean[] array = new boolean[count];
+    for (int i = 0; i < count; ++i) {
+      array[i] = (boolData.getAccessor().get(rowId + i) == 1);
```
Can we move `boolData.getAccessor()` out of the loop if it is a loop invariant?
Ditto for other types (e.g. `getBytes()`).

kiszk commented on sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ArrowColumnVector.java:
```diff
@@ -0,0 +1,502 @@
+
+  @Override
+  public boolean getBoolean(int rowId) {
+    return boolData.getAccessor().get(rowId) == 1;
```
Can we use `nulls`? If so, it would be better to use another name instead of `nulls`.

SparkQA commented:
**[Test build #79668 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79668/testReport)** for PR 18655 at commit [`58cd465`](https://github.com/apache/spark/commit/58cd46506b02800269380f7c8acb5f9825664cad).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79668/
Test PASSed.

kiszk commented:
Good feature, but can we split this PR into smaller PRs for ease of review since it looks large?
For example, since `ArrowColumnVector` is not used in refactored code, this part can be moved to another PR.

BryanCutler commented:
Thanks for this @ueshin.  I agree with @kiszk that it would be easier to review if you can split this into smaller PRs, maybe keep the additional type support separate?  I'm all for refactoring this too, but could you elaborate with some details on why you are refactoring `ColumnWriter` and `ArrowConverters`?  Thanks!

cloud-fan commented:
yea let's put `ArrowColumnVector` and its tests in a new PR and merge that first.

`ArrowWriter` will also be used for pandas UDF, see https://issues.apache.org/jira/browse/SPARK-21190 for more details, so it makes sense to move it to a separated file.

ueshin commented:
Thank you for your comments.
I agree that we should split this into smaller PRs. I'll push another commit to remove `ArrowColumnVector` from this as soon as possible.

ueshin committed:
Revert ArrowColumnVector related implementations to split a pr into multiple prs.

SparkQA commented:
**[Test build #79696 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79696/testReport)** for PR 18655 at commit [`8ffedda`](https://github.com/apache/spark/commit/8ffedda9f05d379d700aef95dca049a751374f87).

SparkQA commented:
**[Test build #79696 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79696/testReport)** for PR 18655 at commit [`8ffedda`](https://github.com/apache/spark/commit/8ffedda9f05d379d700aef95dca049a751374f87).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79696/
Test FAILed.

ueshin commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #79698 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79698/testReport)** for PR 18655 at commit [`8ffedda`](https://github.com/apache/spark/commit/8ffedda9f05d379d700aef95dca049a751374f87).

ueshin commented:
@BryanCutler I'd like to share the motivation of refactoring `ArrowConverters` and `ColumnWriter`.

For `ColumnWriter`, at first I'd like to support complex types like `ArrayType` and `StructType`, so I refactored it based on your `ColumnWriter` implementation. And then I renamed and moved the package so that we can also use it for pandas UDF as @cloud-fan mentioned. As you might see before, I'll introduce `ArrowColumnVector` as a reader for Arrow vectors as well.

For `ArrowConverters`, I thought we can skip the intermediate `ArrowRecordBatch` creation in `ArrowConverters.toPayloadIterator()`. What do you think about that?

Thanks!

SparkQA commented:
**[Test build #79698 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79698/testReport)** for PR 18655 at commit [`8ffedda`](https://github.com/apache/spark/commit/8ffedda9f05d379d700aef95dca049a751374f87).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79698/
Test PASSed.

BryanCutler commented:
> For ArrowConverters, I thought we can skip the intermediate ArrowRecordBatch creation in ArrowConverters.toPayloadIterator(). What do you think about that?

Ok, I see.  By using `ArrowWriter` directly on the root, then the `ArrowFileWriter` can use that same root in creating the Byte array.  So no need to create an intermediate `ArrowRecordBatch`.  That sounds good to me!

> For ColumnWriter, at first I'd like to support complex types like ArrayType and StructType, so I refactored it based on your ColumnWriter implementation. And then I renamed and moved the package 

That's fine, but do they need to be in `o.a.s.sql.execution.vectorized`?  If so, then what's the point of having a `o.a.s.sql.execution.arrow` package if `ArrowUtils` and `ArrowWriter` are not even there?

BryanCutler commented on sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:
```diff
@@ -55,145 +51,55 @@ private[sql] class ArrowPayload private[arrow] (payload: Array[Byte]) extends Se
-      recordsInBatch += 1
+    context.addTaskCompletionListener { _ =>
+      root.close()
+      allocator.close()
```
It seems a little odd to me to tie an iterator to a TaskContext, why not just close resources as soon as the row iterator is consumed?

ueshin replied:
I was worried about memory leak when an exception happens during iterating. In that case, the task will fail before the row iterator is completely consumed.

BryanCutler replied:
Yeah, good point.  What about closing resources in both ways?  So have the listener close for the case that something fails, otherwise once the the row iterator is fully consumed then close immediately.  I'm not really sure at what exact point the task completion listener callback is done, is it dependent on any IO?

BryanCutler reviewed (commented):
Thanks for the PR @ueshin!  I have some concerns about using the TaskContext in the iterator to release resources.  I believe there was a bug fix in Arrow 0.4.1 for the decimal type, and I had planned to look into upgraded to support that type.  Also, with the added type support here, how does it affect the python side and are you planning on adding tests there?

As for the refactoring, if it offers a performance improvement that is great.  However, it seems a little out  of order to me to refactor and move files around to support a SPIP that has not reached a consensus and not been voted on.  Just my thoughts.

ueshin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:
```diff
@@ -55,145 +51,55 @@ private[sql] class ArrowPayload private[arrow] (payload: Array[Byte]) extends Se
+        val writer = new ArrowFileWriter(root, null, Channels.newChannel(out))
+
+        Utils.tryWithSafeFinally {
+          var rowId = 0
```
Thanks! I'll update it.

ueshin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/vectorized/ArrowWriter.scala:
```diff
@@ -0,0 +1,405 @@
+
+  def setNull(): Unit
+  def setValue(input: SpecializedGetters, ordinal: Int): Unit
+  def skip(): Unit
```
This is for the case if the value of the struct type is null.
I believe if the value of the struct type, the fields should have some values for the same row.

ueshin commented on sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala:
```diff
@@ -391,6 +392,85 @@ class ArrowConvertersSuite extends SharedSQLContext with BeforeAndAfterAll {
     collectAndValidate(df, json, "floating_point-double_precision.json")
   }
 
+  ignore("decimal conversion") {
```
Oh, I'm sorry, I should have mentioned it.
It seems like `JsonFileReader` doesn't support DecimalType, so I ignored it for now.
But now I'm thinking that If Arrow 0.4.0 has a bug for the decimal type as you said, should I remove decimal type support from this pr and add support in the following prs?

BryanCutler replied:
That might be true, I haven't looked into it yet.  I can work on adding support on the Arrow side, so I'll try to check on that and see where it stands in the upcoming 0.5 release.

BryanCutler replied:
Arrow integration support for DecimalType isn't slated until v0.6, so it might work but there are no guarantees that a record batch in Java will equal when that batch is read by Python/C++.  Also, we can't test here until the `JsonFileReader` supports it also.  I made the Arrow JIRA here https://issues.apache.org/jira/browse/ARROW-1238

ueshin commented:
I see, I'll move files back to `arrow` package.

ueshin committed:
Use rowCount instead of rowId.

ueshin committed:
Move files back to arrow package.

SparkQA commented:
**[Test build #79737 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79737/testReport)** for PR 18655 at commit [`b5988f9`](https://github.com/apache/spark/commit/b5988f9a223de407b7709f239fca672bb02b60aa).

ueshin committed:
Modify ArrowUtils to avoid deprecated APIs.

SparkQA commented:
**[Test build #79741 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79741/testReport)** for PR 18655 at commit [`a50a271`](https://github.com/apache/spark/commit/a50a271b2738cf25748f2376935d5b30bf4bc3aa).

SparkQA commented:
**[Test build #79737 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79737/testReport)** for PR 18655 at commit [`b5988f9`](https://github.com/apache/spark/commit/b5988f9a223de407b7709f239fca672bb02b60aa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79737/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79741/
Test FAILed.

ueshin commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #79744 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79744/testReport)** for PR 18655 at commit [`a50a271`](https://github.com/apache/spark/commit/a50a271b2738cf25748f2376935d5b30bf4bc3aa).

SparkQA commented:
**[Test build #79744 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79744/testReport)** for PR 18655 at commit [`a50a271`](https://github.com/apache/spark/commit/a50a271b2738cf25748f2376935d5b30bf4bc3aa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79744/
Test PASSed.

ueshin committed:
Modify to close resources also immediately after row iterator is consumed.

SparkQA commented:
**[Test build #79786 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79786/testReport)** for PR 18655 at commit [`7084b38`](https://github.com/apache/spark/commit/7084b388d87c8347b79898827658d7827bf5649d).

SparkQA commented:
**[Test build #79786 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79786/testReport)** for PR 18655 at commit [`7084b38`](https://github.com/apache/spark/commit/7084b388d87c8347b79898827658d7827bf5649d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79786/
Test PASSed.

ueshin committed:
Merge branch 'master' into issues/SPARK-21440

ueshin committed:
Add ArrowWriterSuite.

SparkQA commented:
**[Test build #79798 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79798/testReport)** for PR 18655 at commit [`6fc4da0`](https://github.com/apache/spark/commit/6fc4da05a84ee55ec8fd98c078c16d5671a303cc).

SparkQA commented:
**[Test build #79798 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79798/testReport)** for PR 18655 at commit [`6fc4da0`](https://github.com/apache/spark/commit/6fc4da05a84ee55ec8fd98c078c16d5671a303cc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class ArrowWriterSuite extends SparkFunSuite `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79798/
Test PASSed.

wesm commented:
On DecimalType, I want to point out that we haven't hardened the memory format and integration tests between Java<->C++ within Arrow. It would be great if you could help with this -- we ran into a problem in C++ where we needed an extra sign bit with 16-byte high precision decimals. So we have 3 memory representations:

- 4 byte decimals (low precision)
- 8 byte
- 16 byte plus sign bitmap

What is Spark's internal memory representation? cc @cpcloud

cpcloud commented:
Based on this code: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnType.scala#L429-L547

It looks like there are two types:

1. 8 byte compact decimal that fits in a Java `long`
2. Up to length 16 `Array[byte]` representation (based on the `BINARY`) type.

I think Arrow's Decimal representation in Java is almost identical to this.

Looking at the `BigInteger` Java implementation (which is what `BigDecimal` sits on top of) the sign is carried around in the first `byte` of the array.

wesm commented:
There are a bunch of open JIRAs about decimals in Arrow: https://issues.apache.org/jira/issues/?filter=12334829&jql=project%20%3D%20ARROW%20AND%20status%20in%20(%22In%20Review%22%2C%20Open%2C%20%22In%20Progress%22%2C%20Reopened)%20AND%20text%20~%20%22decimal%22 between these JIRAs and the mailing list if would be good to come up with a game plan for integration tests between Java and C++ (and thus Python) so we can enable Spark to send Python decimals

ueshin commented:
@BryanCutler @wesm @cpcloud Thank you for reviewing this.
If the remaining issue here is only `DecimalType` support, I'd like to separate it from this pr and merge this first to avoid duplicating efforts around writers.
What do you think?

cloud-fan commented:
yes let leave decimal support for folllow-ups

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:
```diff
@@ -77,95 +59,55 @@ private[sql] object ArrowConverters {
-        columnWriters(i).write(row)
-        i += 1
+    context.addTaskCompletionListener { _ =>
+      if (!closed) {
```
do we really need this? I think it's ok to close twice?

ueshin replied:
The `allocator` can be closed twice, but the `root` throws an exception after `allocator` is closed.

cloud-fan replied:
is this a bug in arrow?  cc @BryanCutler

BryanCutler replied:
The root just releases the buffers from the FieldVectors, so I would think it should be able to handle being closed twice.  I'll check tomorrow if seems reasonable.

BryanCutler replied:
I filed https://issues.apache.org/jira/browse/ARROW-1283 to fix this.  For now, it looks like we need this.

ueshin committed:
Merge branch 'master' into issues/SPARK-21440

ueshin committed:
Remove DecimalType support for now.

BryanCutler commented:
+1 on holding off for `DecimalType` support

SparkQA commented:
**[Test build #79955 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79955/testReport)** for PR 18655 at commit [`5bbb46f`](https://github.com/apache/spark/commit/5bbb46f55a90bae3008ee374ef2c0349a3489c09).

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala:
```diff
@@ -0,0 +1,383 @@
+
+  def writeSkip(): Unit = {
+    skip()
+    count += 1
```
For skipping purpose, is it enough to just do `count += 1`? e.g. `vector.set(1, v1); vector.set(3, v3)`, value 2 is skipped.

ueshin replied:
Basically, yes, it's enough except for `StructType`, but should we set null bit to `1` for skipped value?

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowWriterSuite.scala:
```diff
@@ -0,0 +1,247 @@
+      val reader = new ArrowColumnVector(writer.root.getFieldVectors().get(0))
+      data.zipWithIndex.foreach {
+        case (null, rowId) => assert(reader.isNullAt(rowId))
+        case (datum, rowId) => assert(get(reader, rowId) === datum)
```
we can do something like
```
dt match {
 case BooleanType => reader.getBoolean(rowid)
 case IntegerType => ...
 ...
}
```
Then the caller side doesn't need to pass in a `get`

ueshin replied:
Thanks, I'll update it.

HyukjinKwon commented on sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala:
```diff
@@ -857,6 +857,449 @@ class ArrowConvertersSuite extends SharedSQLContext with BeforeAndAfterAll {
+    val a_arr = Seq(Seq(1, 2), Seq(3, 4), Seq(), Seq(5))
+    val b_arr = Seq(Some(Seq(1, 2)), None, Some(Seq()), None)
+    val c_arr = Seq(Seq(Some(1), Some(2)), Seq(Some(3), None), Seq(), Seq(Some(5)))
+    val d_arr = Seq(Seq(Seq(1, 2)), Seq(Seq(3), Seq()), Seq(), Seq(Seq(5)))
```
How about camelCase naming?

ueshin replied:
Thanks, I'll modify it.

ueshin committed:
Fix skip() for StructType.

ueshin commented on sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowWriter.scala:
```diff
@@ -0,0 +1,383 @@
+
+  private def createFieldWriter(vector: ValueVector): ArrowFieldWriter = {
+    val field = vector.getField()
+    ArrowUtils.fromArrowField(field) match {
```
Thanks, I'll modify it.

ueshin committed:
Address comments.

SparkQA commented:
**[Test build #79957 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79957/testReport)** for PR 18655 at commit [`19f3973`](https://github.com/apache/spark/commit/19f3973c4acf1b05ae51c338481d975cebf66a98).

SparkQA commented:
**[Test build #79955 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79955/testReport)** for PR 18655 at commit [`5bbb46f`](https://github.com/apache/spark/commit/5bbb46f55a90bae3008ee374ef2c0349a3489c09).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79955/
Test PASSed.

SparkQA commented:
**[Test build #79957 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79957/testReport)** for PR 18655 at commit [`19f3973`](https://github.com/apache/spark/commit/19f3973c4acf1b05ae51c338481d975cebf66a98).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79957/
Test FAILed.

ueshin commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #79960 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79960/testReport)** for PR 18655 at commit [`19f3973`](https://github.com/apache/spark/commit/19f3973c4acf1b05ae51c338481d975cebf66a98).

SparkQA commented:
**[Test build #79960 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79960/testReport)** for PR 18655 at commit [`19f3973`](https://github.com/apache/spark/commit/19f3973c4acf1b05ae51c338481d975cebf66a98).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79960/
Test PASSed.

ueshin committed:
Modify skip semantic.

SparkQA commented:
**[Test build #79995 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79995/testReport)** for PR 18655 at commit [`0bac10d`](https://github.com/apache/spark/commit/0bac10d95637c1afa632210b5feca079a61a35d2).

ueshin committed:
Inline writeNull().

SparkQA commented:
**[Test build #79996 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79996/testReport)** for PR 18655 at commit [`b85dc23`](https://github.com/apache/spark/commit/b85dc231d05f5e1a1a3d8b0bcbc778b85d83c533).

SparkQA commented:
**[Test build #79995 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79995/testReport)** for PR 18655 at commit [`0bac10d`](https://github.com/apache/spark/commit/0bac10d95637c1afa632210b5feca079a61a35d2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79995/
Test PASSed.

SparkQA commented:
**[Test build #79996 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/79996/testReport)** for PR 18655 at commit [`b85dc23`](https://github.com/apache/spark/commit/b85dc231d05f5e1a1a3d8b0bcbc778b85d83c533).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/79996/
Test PASSed.

cloud-fan commented:
LGTM, merging to master!

ueshin commented:
@BryanCutler @wesm @cpcloud I filed a JIRA issue for decimal type support [SPARK-21552](https://issues.apache.org/jira/browse/SPARK-21552) and sent a pr for it as WIP #18754.
Let's move on there for discussing decimal type support.
